unmarshal.go

```diff
@@ -0,0 +1,298 @@
+package apiclient
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/golang/protobuf/proto"
+	"reflect"
+	"strings"
+)
+
+// Unmarshal unmarshals a byte array into a predefined proto.Message.
+func Unmarshal(b []byte, pm proto.Message) error {
+	jsonMap := make(map[string]interface{})
+	if err := json.Unmarshal(b, &jsonMap); err != nil {
+		return err
+	}
+	grrValue, hasValue := jsonMap["value"]
+	grrType, hasType := jsonMap["type"].(string)
+	if hasValue && hasType {
+		return unmarshalTypeValue(pm, grrType, grrValue)
+	} else {
+		return unmarshalMap(pm, jsonMap)
+	}
+	return nil
+}
+
+// unmarshalTypeValue unmarshals a value to a predefined object. The
+// value has to be a pointer to bool, int, float, string or struct.
+// If the value is a pointer to struct, the grrType is used to unmarshal
+// into proto.Message.
+func unmarshalTypeValue(obj interface{}, grrType string, grrValue interface{}) error {
+
+	protoType := reflect.TypeOf(obj)
+	protoValue := reflect.ValueOf(obj).Elem()
+
+	/*
+	 * Check if the object brings an unmarshaler with it and use it
+	 */
+	if protoType.Implements(reflect.TypeOf((*json.Unmarshaler)(nil)).Elem()) {
+		unmarshaler := reflect.New(protoType.Elem()).Interface().(json.Unmarshaler)
+		grrBytes, err := json.Marshal(grrValue)
+		if err != nil {
+			return err
+		}
+		if err = unmarshaler.UnmarshalJSON(grrBytes); err != nil {
+			return err
+		}
+		protoValue.Set(reflect.ValueOf(unmarshaler).Elem())
+		return nil
+	}
+
+	/*
+	 * Else check what kind of interface the
+	 * object to be written is
+	 */
+	switch protoType.Elem().Kind() {
```

Comment:
It should be possible to do this using a type switch instead of going through reflect.